### PR TITLE
FIX-#5598: make `PyArrowDataset.files` work for `3.0.0 <= pyarrow < 8.0.0`

### LINKED ISSUE #5598: `PyArrowDataset.files` uses `ParquetDataset.files` method that exists only for `pyarrow>=8.0.0`
We should update the minimum version of this dependency.

https://github.com/modin-project/modin/blob/5d1fe8b1bb789e30292bb39da45f99f71f48bfc6/modin/core/io/column_stores/parquet_dispatcher.py#L218

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

We use an interface that only appears with `pyarrow>=8.0.0` version. https://arrow.apache.org/docs/8.0/python/generated/pyarrow.parquet.ParquetDataset.html#pyarrow.parquet.ParquetDataset.files
```python
  File "C:\projects\modin\modin\core\io\column_stores\parquet_dispatcher.py", line 218, in files
    self._files = self._get_files(self.dataset.files)
AttributeError: '_ParquetDatasetV2' object has no attribute 'files'
```

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5598 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by anmyachev** [comment]
@mvashishtha what do you think about this change?

**Comment by YarShev** [comment]
If we pin pyarrow to >=8.0.0, we will lose interoperability with Ray dataset (sed deps for it [here](https://github.com/ray-project/ray/blob/f2b6a6baced3f252c0e04249224c870299a4998c/python/setup.py#L220-L225)) and probably some other libraries. I would say that we should leave the pyarrow installation as is for now.

**Comment by anmyachev** [comment]
> If we pin pyarrow to >=8.0.0, we will lose interoperability with Ray dataset (see deps for it [here](https://github.com/ray-project/ray/blob/f2b6a6baced3f252c0e04249224c870299a4998c/python/setup.py#L220-L225)) and probably some other libraries. I would say that we should leave the pyarrow installation as is for now.

Good point.

@pyrito if we do not change the version of `pyarrow`, then we need to rewrite the implementation of https://github.com/modin-project/modin/blob/91d6825c93ea58d790352229d798ff7cbecf50dd/modin/core/io/column_stores/parquet_dispatcher.py#L216 function to use pyarrow interface that available from `4.0.1`. Can you do this?

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 1]
```diff
@@ -215,7 +215,12 @@ def row_groups_per_file(self):
     @property
     def files(self):
         if self._files is None:
-            self._files = self._get_files(self.dataset.files)
+            try:
+                files = self.dataset.files
+            except AttributeError:
+                # compatibility with 4.0.1 <= pyarrow < 8.0.0
+                files = self.dataset._dataset.files
```
Comment (anmyachev) [inline modin/core/io/column_stores/parquet_dispatcher.py:222]: tested the code locally for the next `pyarrow` versions: 4.0.1, 5.0.0, 6.0.0, 7.0.0, 8.0.0

**Comment by anmyachev** [comment]
> If we pin pyarrow to >=8.0.0, we will lose interoperability with Ray dataset (see deps for it [here](https://github.com/ray-project/ray/blob/f2b6a6baced3f252c0e04249224c870299a4998c/python/setup.py#L220-L225)) and probably some other libraries. I would say that we should leave the pyarrow installation as is for now.

@YarShev I've implemented another approach. Could you take a look?

**Comment by pyrito** [comment]
@anmyachev I apologize for not responding to this earlier. I think it got lost in my notifications. Were you able to get this sorted?

**Comment by dchigarev** [comment]
The changes look good, but don't we need to pin pyarrow version to `>=4.0.1` then? 

BTW, why 4.0.1? Shouldn't it work on older versions? I doubt that they added new functionality in a patch release, I bet this `._dataset.files` thing should work at least at 4.0.0.

**Comment by anmyachev** [comment]
> The changes look good, but don't we need to pin pyarrow version to `>=4.0.1` then?

Let's not do this, I just checked that reading (`read-parquet`) works with these versions.

> BTW, why 4.0.1? Shouldn't it work on older versions? I doubt that they added new functionality in a patch release, I bet this `._dataset.files` thing should work at least at 4.0.0.

You're right, moreover, this code works for `3.0.0` as well, possibly for older ones too, but I haven't tested it. I took release 4.0.1 as the minimum version simply because it was not working for the user.

**Comment by anmyachev** [comment]
> Were you able to get this sorted?

@pyrito looks like that

**Comment by anmyachev** [comment]
@pyrito are you ok with the changes?

**Review by dchigarev — APPROVED** [review]

**Pull request merged**